Update stb_vorbis to v1.22

ccawley2011 opened the pull request:
All tests pass, however `test_fuzzer_depack_oxm_truncated` takes longer than expected. Some of the fixes previously made to version 0.99996 may need to be reapplied.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -232,29 +623,30 @@
 #endif
 
 
+#if 0
+#include <crtdbg.h>
+#define CHECK(f)   _CrtIsValidHeapPointer(f->channel_buffers[1])
+#else
+#define CHECK(f)   ((void) 0)
+#endif
+
 #define MAX_BLOCKSIZE_LOG  13   // from specification
 #define MAX_BLOCKSIZE      (1 << MAX_BLOCKSIZE_LOG)
 
 
-#if 0
 typedef unsigned char  uint8;
 typedef   signed char   int8;
 typedef unsigned short uint16;
 typedef   signed short  int16;
 typedef unsigned int   uint32;
 typedef   signed int    int32;
-#endif
 
```
This looks like a libxmp change that got clobbered.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -710,16 +1078,12 @@ static int compute_codewords(Codebook *c, uint8 *len, int n, uint32 *values)
    memset(available, 0, sizeof(available));
    // find the first entry
    for (k=0; k < n; ++k) if (len[k] < NO_CODE) break;
-
-   // sanity check
-   if (k == n) return (c->sorted_entries == 0);
-   if (len[k] >= 32) return FALSE;
-
+   if (k == n) { assert(c->sorted_entries == 0); return TRUE; }
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -734,19 +1098,15 @@ static int compute_codewords(Codebook *c, uint8 *len, int n, uint32 *values)
       // more than one free leaf at a given level, isn't totally
       // trivial to prove, but it seems true and the assert never
       // fires, so!
-
-      /* Sanity check */
-      if (z >= 32) {
-        return FALSE;
-      }
-
       while (z > 0 && !available[z]) --z;
-      if (z == 0) { /*assert(0);*/ return FALSE; }
+      if (z == 0) { return FALSE; }
       res = available[z];
+      assert(z >= 0 && z < 32);
```
clobbered bugfix, (Also, asserts are a really poor excuse for bounds checks on arbitrary data.)

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -789,7 +1155,7 @@ static int uint32_compare(const void *p, const void *q)
 
 static int include_in_sort(Codebook *c, uint8 len)
 {
-   if (c->sparse) { return (len != NO_CODE); }
+   if (c->sparse) { assert(len != NO_CODE); return TRUE; }
```
more asserts where upstream should be doing bounds checks.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -902,14 +1266,8 @@ static void compute_bitreverse(int n, uint16 *rev)
 {
    int ld = ilog(n) - 1; // ilog is off-by-one from normal definitions
    int i, n8 = n >> 3;
-   for (i=0; i < n8; ++i) {
-      /* CID 128660 (#1 of 1): Bad bit shift operation (BAD_SHIFT)
-       * large_shift: right shifting by more than 31 bits has undefined
-       * behavior. The shift amount, 32 - ld + 3, is 36.
-       */
-      int s = 32-ld+3;
-      rev[i] = s > 31 ? 0 : (bit_reverse(i) >> s) << 2;
-   }
+   for (i=0; i < n8; ++i)
+      rev[i] = (bit_reverse(i) >> (32-ld+3)) << 2;
```
clobbered bugfix.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1243,23 +1606,14 @@ static uint32 get_bits(vorb *f, int n)
          f->valid_bits += 8;
       }
    }
-   if (f->valid_bits < 0) return 0;
+
+   assert(f->valid_bits >= n);
```
🙄

ccawley2011 replied:
This appears to have been a deliberate change upstream: https://github.com/nothings/stb/pull/935

AliceLR replied:
Got it—I didn't check the surrounding code, the check above that is the one that actually matters.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1283,15 +1637,17 @@ enum
 {
    VORBIS_packet_id = 1,
    VORBIS_packet_comment = 3,
-   VORBIS_packet_setup = 5,
+   VORBIS_packet_setup = 5
 };
 
 static int codebook_decode_scalar_raw(vorb *f, Codebook *c)
 {
    int i;
    prep_huffman(f);
 
-   assert(c->sorted_codewords || c->codewords);
+   if (c->codewords == NULL && c->sorted_codewords == NULL)
+      return -1;
+
```
🙄

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1557,7 +1884,8 @@ static int codebook_decode_deinterleave_repeat(vorb *f, Codebook *c, float **out
          for (i=0; i < effective; ++i) {
             int off = (z / div) % c->lookup_values;
             float val = CODEBOOK_ELEMENT_FAST(c,off) + last;
-            outputs[c_inter][p_inter] += val;
+            if (outputs[c_inter])
+               outputs[c_inter][p_inter] += val;
```
Come on...

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1569,14 +1897,16 @@ static int codebook_decode_deinterleave_repeat(vorb *f, Codebook *c, float **out
          if (c->sequence_p) {
             for (i=0; i < effective; ++i) {
                float val = CODEBOOK_ELEMENT_FAST(c,z+i) + last;
-               outputs[c_inter][p_inter] += val;
+               if (outputs[c_inter])
+                  outputs[c_inter][p_inter] += val;
```
another NULL dereference

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1569,14 +1897,16 @@ static int codebook_decode_deinterleave_repeat(vorb *f, Codebook *c, float **out
          if (c->sequence_p) {
             for (i=0; i < effective; ++i) {
                float val = CODEBOOK_ELEMENT_FAST(c,z+i) + last;
-               outputs[c_inter][p_inter] += val;
+               if (outputs[c_inter])
+                  outputs[c_inter][p_inter] += val;
                if (++c_inter == ch) { c_inter = 0; ++p_inter; }
                last = val;
             }
          } else {
             for (i=0; i < effective; ++i) {
                float val = CODEBOOK_ELEMENT_FAST(c,z+i) + last;
-               outputs[c_inter][p_inter] += val;
+               if (outputs[c_inter])
+                  outputs[c_inter][p_inter] += val;
```
another NULL dereference

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1774,11 +2028,6 @@ static __forceinline int draw_line(float *output, int x0, int y0, int x1, int y1
    int err = 0;
    int sy;
 
-   /* Sanity check */
-   if (adx == 0) {
-	return -1;
-   }
-
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1804,31 +2053,18 @@ static __forceinline int draw_line(float *output, int x0, int y0, int x1, int y1
    ady -= abs(base) * adx;
    if (x1 > n) x1 = n;
-
-   /* Sanity check */
-   if (x >= n * 2 || y >= 256) {
-     return -1;
-   }
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -1804,31 +2053,18 @@ static __forceinline int draw_line(float *output, int x0, int y0, int x1, int y1
-         y += base;
-      }
-
-      /* Sanity check */
-      if (y >= 256) {
-        return -1;
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3070,10 +3255,6 @@ static int vorbis_decode_packet_rest(vorb *f, int *len, Mode *m, int left_start,
             }
 
 #ifdef STB_VORBIS_NO_DEFER_FLOOR
-	    /* Sanity check */
-	    if (n >= f->blocksize_1)
-		return FALSE;
-
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3128,19 +3305,14 @@ static int vorbis_decode_packet_rest(vorb *f, int *len, Mode *m, int left_start,
          }
       }
       r = map->submap_residue[i];
-      t = f->residue_types[r];
-      if (!decode_residue(f, residue_buffers, ch, n2, r, do_not_decode)) {
-         return FALSE;
-      }
```
Not sure if this one was a bugfix bug it looks suspicious.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3321,16 +3494,15 @@ static int vorbis_finish_frame(stb_vorbis *f, int len, int left, int right)
 
 static int vorbis_pump_first_frame(stb_vorbis *f)
 {
-   int len, right, left;
-   if (vorbis_decode_packet(f, &len, &left, &right)) {
-      if (vorbis_finish_frame(f, len, left, right) < 0)
-         return -1;
-   }
-   return 0;
+   int len, right, left, res;
+   res = vorbis_decode_packet(f, &len, &left, &right);
+   if (res)
+      vorbis_finish_frame(f, len, left, right);
+   return res;
```
What's going on with the return values here?

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3472,17 +3699,16 @@ static int start_decoder(vorb *f)
    // codebooks
 
    f->codebook_count = get_bits(f,8) + 1;
-   if (f->valid_bits < 0)                           return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3494,7 +3720,7 @@ static int start_decoder(vorb *f)
       ordered = get_bits(f,1);
       c->sparse = ordered ? 0 : get_bits(f,1);
 
-      if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix (the new fix below is not equivalent).

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3509,37 +3735,33 @@ static int start_decoder(vorb *f)
          while (current_entry < c->entries) {
             int limit = c->entries - current_entry;
             int n = get_bits(f, ilog(limit));
-            if (f->valid_bits < 0) { res = error(f, VORBIS_unexpected_eof); break; }
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3509,37 +3735,33 @@ static int start_decoder(vorb *f)
       } else {
          for (j=0; j < c->entries; ++j) {
             int present = c->sparse ? get_bits(f,1) : 1;
-            if (f->valid_bits < 0) { res = error(f, VORBIS_unexpected_eof); break; }
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3509,37 +3735,33 @@ static int start_decoder(vorb *f)
             }
          }
       }
-      if (res != TRUE) {
-         if (c->sparse) {
-            setup_temp_free(f, lengths, c->entries);
-         }
-         return res;
-      }
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3616,98 +3839,76 @@ static int start_decoder(vorb *f)
-                         #endif
-                  div *= c->lookup_values;
-
-                  /* Sanity check */
-                  if (div == 0) {
-                    free(mults);
-                    return error(f, VORBIS_invalid_setup);
```
this might be a clobbered bugfix.

AliceLR replied:
OK, I think I know what's going on here—the new check at 3924 should prevent this case from occurring (and even if it did occur, the `% c->lookup_values` at 3952 would SIGFPE first anyway). So this change looks good.

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3720,8 +3921,8 @@ static int start_decoder(vorb *f)
 
    // Floors
    f->floor_count = get_bits(f, 6)+1;
-   if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3749,7 +3950,6 @@ static int start_decoder(vorb *f)
          for (j=0; j <= max_class; ++j) {
             g->class_dimensions[j] = get_bits(f, 3)+1;
             g->class_subclasses[j] = get_bits(f, 2);
-            if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3794,51 +3997,40 @@ static int start_decoder(vorb *f)
 
    // Residue
    f->residue_count = get_bits(f, 6)+1;
-   if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3794,51 +3997,40 @@ static int start_decoder(vorb *f)
       r->part_size = get_bits(f,24)+1;
       r->classifications = get_bits(f,6)+1;
       r->classbook = get_bits(f,8);
-      if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3794,51 +3997,40 @@ static int start_decoder(vorb *f)
             high_bits = get_bits(f,5);
          residue_cascade[j] = high_bits*8 + low_bits;
       }
-      if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3794,51 +3997,40 @@ static int start_decoder(vorb *f)
          for (k=0; k < 8; ++k) {
             if (residue_cascade[j] & (1 << k)) {
                r->residue_books[j][k] = get_bits(f, 8);
-               if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3856,25 +4049,27 @@ static int start_decoder(vorb *f)
    }
 
    f->mapping_count = get_bits(f,6)+1;
-   if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3856,25 +4049,27 @@ static int start_decoder(vorb *f)
          for (k=0; k < m->coupling_steps; ++k) {
-            m->chan[k].magnitude = get_bits(f, ilog(f->channels)-1);
-            m->chan[k].angle = get_bits(f, ilog(f->channels)-1);
-            if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -3911,7 +4106,6 @@ static int start_decoder(vorb *f)
       m->windowtype = get_bits(f,16);
       m->transformtype = get_bits(f,16);
       m->mapping = get_bits(f,8);
-      if (f->valid_bits < 0)                  return error(f, VORBIS_unexpected_eof);
```
clobbered bugfix

AliceLR commented on src/depackers/stb_vorbis.c:
```diff
@@ -4266,30 +4489,24 @@ int stb_vorbis_decode_frame_pushdata(
 
    // success!
    len = vorbis_finish_frame(f, len, left, right);
-
-   /* Sanity check */
-   if (len < 0) {
-      return -1;
-   }
-
```
clobbered bugfix?

AliceLR reviewed (changes requested):
This is incomplete (the diff turns into a mess around ~5000 lines in) but I tried to note all of the bugfixes this broke, including several that were added to fix slow loads. I'm not 100% sure all of these were libxmp fixes but most of the missing checks in `start_decoder` (particularly on `f->valid_bits`) were ones I added.

sezero commented on src/depackers/oxm.c:
```diff
@@ -20,7 +20,7 @@
  */
 
-#include "vorbis.h"
+#include "stb_vorbis.h"
```
IMO, we should keep the name as it is w/o `stb_` prefix.

Will look at the rest later.

AliceLR replied:
I second leaving the filenames the same for now. This would simplify the patch a little, and this file may end up getting renamed/moved to `src/loaders/` in the near future anyway (Claudio greenlit merging the OXM depacker into the sample code today in email).

sezero commented:
I usually apply the following patches in other projects:
- tweak `alloca()` discovery: [0001_patch.txt](https://github.com/libxmp/libxmp/files/6760604/0001_patch.txt)
- tweak `forceinline` definition: [0002_patch.txt](https://github.com/libxmp/libxmp/files/6760605/0002_patch.txt)
- remove unnecessary comments support: [0003_patch.txt](https://github.com/libxmp/libxmp/files/6760606/0003_patch.txt)
- rename `BUFFER_SIZE` macro to `STB_BUFFER_SIZE`: [0004_patch.txt](https://github.com/libxmp/libxmp/files/6760607/0004_patch.txt)

ccawley2011 commented:
> This is incomplete (the diff turns into a mess around ~5000 lines in) but I tried to note all of the bugfixes this broke, including several that were added to fix slow loads. I'm not 100% sure all of these were libxmp fixes but most of the missing checks in `start_decoder` (particularly on `f->valid_bits`) were ones I added.
> 
> edit: deleted some comments I noted that were incorrect...
> edit: since there are so many of these, I'll try to sift through them and figure out which ones actually need to be upstreamed.

I've attempted to restore as much as possible, but I'm a bit out of my depth here, so I'm not 100% sure how much is actually necessary with the new version.

ccawley2011 commented:
> I usually apply the following patches in other projects:
> 
> * tweak `alloca()` discovery: [0001_patch.txt](https://github.com/libxmp/libxmp/files/6760604/0001_patch.txt)
> * tweak `forceinline` definition: [0002_patch.txt](https://github.com/libxmp/libxmp/files/6760605/0002_patch.txt)
> * remove unnecessary comments support: [0003_patch.txt](https://github.com/libxmp/libxmp/files/6760606/0003_patch.txt)
> * rename `BUFFER_SIZE` macro to `STB_BUFFER_SIZE`: [0004_patch.txt](https://github.com/libxmp/libxmp/files/6760607/0004_patch.txt)

I've applied the 1st, 2nd and 4th patches. For the 3rd patch, could you explain the benefit of this patch?

sezero commented:
> > * remove unnecessary comments support: [0003_patch.txt](https://github.com/libxmp/libxmp/files/6760606/0003_patch.txt)
> 
> I've applied the 1st, 2nd and 4th patches. For the 3rd patch, could you explain the benefit of this patch?

User comments support, even the vendor strings aren't needed. Less malloc'ing, less computing.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -4053,6 +4126,12 @@ static int start_decoder(vorb *f)
+
+      /* Sanity check */
+      if (r->classbook >= f->codebook_count) {
+         return error(f, VORBIS_invalid_setup);
```
It looks like this same check has been added in mainstream, but above around line 4107

AliceLR commented:
> I've attempted to restore as much as possible, but I'm a bit out of my depth here, so I'm not 100% sure how much is actually necessary with the new version.

Thanks for going through those. I'm not really sure what all of the things stb-vorbis does are either—a lot of those checks I had to add trial and error, since there are ~a lot~ of places that are able to get stuck in very slow loops when `EOF` is treated as valid data. Will rereview soon.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -3809,10 +4095,12 @@ static int start_decoder(vorb *f)
       r->classifications = get_bits(f,6)+1;
       r->classbook = get_bits(f,8);
       if (f->valid_bits < 0) return error(f, VORBIS_unexpected_eof);
+      if (r->classbook >= f->codebook_count) return error(f, VORBIS_invalid_setup);
```
This new check seems to supersede our old version at line 4127?

sezero commented:
Some of libxmp customizations are lost: many (if not all) of them can be added
to the new tiny vorbis.h:
- We used to `#define STB_VORBIS_NO_PUSHDATA_API` (see around line 423)
- We used to `#define STB_VORBIS_NO_STDIO` (see around line 432)
- We used to `#define STB_VORBIS_MAX_CHANNELS 1` instead of 16, and also
  guarded certain code by `#if STB_VORBIS_MAX_CHANNELS > 1`, such as all of
  `codebook_decode_deinterleave_repeat` and `compute_[stereo_]samples`, and
  others, or parts of `decode_residue` and others.
- We used to `#define STB_VORBIS_NO_ALLOC_BUFFER`, (looks like a libxmp
  invention) and guarded certain code with it, such as `temp_alloc` & co (see around
  line 918 which has changes.) What to do with that...
- We had renamed crc functions, e.g. `crc32_init` to `libxmp_crc32_init_A`,
  `crc32_update` to `libxmp_crc32_update`
- We had `return 0;/* silence pesky compiler warnings */` at the ends of `get8()`
  and `getn()`
- We used to `#if 0`-out `set_file_offset()` and `ogg_page_header[]`

There maybe others..

sezero commented:
I brought back some of the things I mentioned above, and also macroized
the comments support. Patch attached (supersedes patch no.3 I suggested
before):  [0005_patch.txt](https://github.com/libxmp/libxmp/files/6767328/0005_patch.txt)

- The `STB_VORBIS_MAX_CHANNELS` and `STB_VORBIS_NO_ALLOC_BUFFER` issues I
  mentioned above remain (if they are issues..)
- The renaming of crc functions doesn't seem necessary to me in the new version.
- The new version seems to need `set_file_offset()` and `ogg_page_header[]`, so
  can't comment them out, as far as I can see.
- More issues may be there.

ccawley2011 commented:
> I brought back some of the things I mentioned above, and also macroized
> the comments support. Patch attached (supersedes patch no.3 I suggested
> before): [0005_patch.txt](https://github.com/libxmp/libxmp/files/6767328/0005_patch.txt)

I've applied these changes.

Can any of the changes in this PR be upstreamed as is, or is more work needed first?

sezero commented:
> Can any of the changes in this PR be upstreamed as is, or is more work needed first?

They can be I guess, but we should finalize our side first IMO.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -3276,12 +3530,7 @@ static int vorbis_finish_frame(stb_vorbis *f, int len, int left, int right)
-         return -1;
-      }
-
+      if (w == NULL) return 0;
```
The mainstream version was added by https://github.com/nothings/stb/commit/98fdfc6df88b1e34a736d5e126e6c8139c8de1a6 but it doesn't fail like we do:
We rely on a negative return from `vorbis_finish_frame` to fail in our sanity checks.

`vorbis_pump_first_frame` is affected too: down below it returns 0 if it receives
non-zero from it, should  it not fail?

AliceLR replied:
I dug through the code a bit and I think any semantic difference between the upstream version and ours seems irrelevant here—`f->window[]` values should only be `NULL` if early stages of setup failed, weren't called, or if some memory was clobbered.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -2135,33 +2380,27 @@ void inverse_mdct_slow(float *buffer, int n, vorb *f, int blocktype)
 // transform to use a slow dct-iv; this is STILL basically trivial,
 // but only requires half as many ops
-#if 0
-static void dct_iv_slow(float *buffer, int n)
```
All `dct_iv_slow` and `inverse_mdct_slow` should be static

AliceLR replied:
This is all debug deadcode and is irrelevant for libxmp IMO.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -741,12 +1121,14 @@ static int compute_codewords(Codebook *c, uint8 *len, int n, uint32 *values)
-      if (z == 0) { /*assert(0);*/ return FALSE; }
+      if (z == 0) { return FALSE; }
       res = available[z];
+      assert(z >= 0 && z < 32);
```
This assert is wrong: at the very least it be a line above.
Also, why not return failure and assert() instead?

AliceLR replied:
The better question is why this assert exists at all: both failure cases are impossible. Line 1119 prevents the >=32 case and lines 1123+1124 prevent the <0 case.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -741,12 +1121,14 @@ static int compute_codewords(Codebook *c, uint8 *len, int n, uint32 *values)
-      // propogate availability up the tree
+      // propagate availability up the tree
       if (z != len[i]) {
+         assert(len[i] < 32);
```
why not return failure somehow and assert() instead?

AliceLR replied:
This seems like a duplicate of the one below, marking resolved...

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -741,12 +1121,14 @@ static int compute_codewords(Codebook *c, uint8 *len, int n, uint32 *values)
       if (z != len[i]) {
+         assert(len[i] < 32);
          for (y=len[i]; y > z; --y) {
             assert(available[y] == 0);
```
why not return failure somehow and assert() instead?
(yeah, this was here before the new version..)

AliceLR replied:
It does seem like this line of code should be an actual check and not an assert. I'll dig a little bit and see if there's a reason why it isn't...

AliceLR replied:
OK, the answer is at line 3834: the lengths values are initialized with `get_bits(f, 5) + 1` and a value of 32 is considered an error case. So I think this is technically fine.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -2200,13 +2438,13 @@ extern void mdct_backward(mdct_lookup *init, float *in, float *out);
 mdct_lookup M1,M2;
 
-static void inverse_mdct(float *buffer, int n, vorb *f, int blocktype)
+void inverse_mdct(float *buffer, int n, vorb *f, int blocktype)
```
we should keep it `static`

AliceLR replied:
This is only enabled if `LIBVORBIS_MDCT` is defined, which it doesn't look like libxmp ever used. The implementation of `inverse_mdct` that libxmp uses is static and defined at 2683.

sezero commented on src/depackers/stb_vorbis.c:
```diff
@@ -5230,50 +5517,45 @@ int stb_vorbis_decode_filename(char *filename, int *channels, short **output)
+#endif // STB_VORBIS_NO_INTEGER_CONVERSION
 
-#if 0
 int stb_vorbis_get_samples_float_interleaved(stb_vorbis *f, int channels, float *buffer, int num_floats)
```
`stb_vorbis_get_samples_float_interleaved` and `stb_vorbis_get_samples_float`
aren't used by us, hence they were commented out. Maybe should do that again?

AliceLR replied:
Seconded—we don't use these.

AliceLR commented:
After looking this over a couple of times I think I don't have any further complaints about missing bugfixes. Things that should still probably be handled before this is merged:

* IMO keep the old filenames for now to simplify things.
* Some functions libxmp doesn't use that were dummied before could be dummied out again, but I don't think they're going to contribute much to increased code size.

AliceLR commented:
Re: code size, this is the MinGW64 output after `strip --strip-unneeded`:
```
$ l vorbis.lo stb_vorbis.lo
-rw-r--r-- 1 megazeux None 46259 Jul  9 18:21 stb_vorbis.lo
-rw-r--r-- 1 megazeux None 30439 Jul  9 18:22 vorbis.lo
```
So this update is about a +16k binary size increase currently (still considerably smaller than `libvorbis.a`...).

ccawley2011 commented:
> After looking this over a couple of times I think I don't have any further complaints about missing bugfixes. Things that should still probably be handled before this is merged:
> 
> * IMO keep the old filenames for now to simplify things.
> * Some functions libxmp doesn't use that were dummied before could be dummied out again, but I don't think they're going to contribute much to increased code size.

Both of these things are done, and this PR now includes the changes from v1.22.

AliceLR reviewed (approved):
I'm going to go ahead and approve this.

There are numerous -Wmaybe-uninitialized warnings when I build this with MINGW64. This patch fixes them.
```diff
diff --git a/src/depackers/vorbis.c b/src/depackers/vorbis.c
index ed8a82ba..834bca5e 100644
--- a/src/depackers/vorbis.c
+++ b/src/depackers/vorbis.c
@@ -4782,10 +4782,12 @@ static int get_seek_page_info(stb_vorbis *f, ProbedPage *z)
    z->page_start = stb_vorbis_get_file_offset(f);
 
    // parse the header
-   getn(f, header, 27);
+   if (!getn(f, header, 27))
+      return 0;
    if (header[0] != 'O' || header[1] != 'g' || header[2] != 'g' || header[3] != 'S')
       return 0;
-   getn(f, lacing, header[26]);
+   if (!getn(f, lacing, header[26]))
+      return 0;
 
    // determine the length of the payload
    len = 0;
```

There's also the matter of the old max channels=1 hacks, the removal of which is probably a not-insignificant portion of the code size increase. I don't think this needs to be reimplemented unless someone can confirm that OggMod and MO3 modules are completely incapable of OGG encoding stereo samples, and I definitely don't think it should be a blocker.

sezero commented:
Go ahead and apply.

For future reference, other than the points already mentioned: Should we
build with `-DNDEBUG`? Or, should we change the `assert()`s to error-returns
or something?

AliceLR commented:
> Go ahead and apply.
> 
> For future reference, other than the points already mentioned: Should we
> build with `-DNDEBUG`? Or, should we change the `assert()`s to error-returns
> or something?

As I commented above, the asserts are mostly worthless. It's probably OK to build with `-DNDEBUG` in cases where `-DDEBUG` hasn't also been defined. That belongs in a different branch IMO.
